### ISSUE TITLE
Sublcasses can chang behaviour of parent class (Contract)

### ISSUE THREAD

**Issue by greyblake** [dry-rb/dry-validation#594]
### Steps to reproduce

Considering we have the following code snippet

```ruby
class EmailContract < Dry::Validation::Contract
  params do
    required(:email).filled(:string)
  end
end

class LoginContract < EmailContract
  params do
    required(:password).filled(:string)
  end
end

email_contract = EmailContract.new
result = email_contract.call(email: 'foo@bar.com')
pp result
```

Output (result with errors: password is missing):
```ruby
#<Dry::Validation::Result{:email=>"foo@bar.com"} errors={:password=>["is missing"]}>
```

### Insights

It looks like `LoginContract` and `EmailContract` due to inheritance share some common state which is mutated by defining params schemas for `EmailContract`.

### Expected

I would expect one of the following:
* Subclasses (`LoginContract`) MUST NOT change behaviour of a parent class ( `EmailContract`)
* If inheritance is forbidden by design that it would be meaningful to raise a noisy explaining exception.

### Environment

OS: `Linux dc58b674a4e0 5.0.0-27-generic #28-Ubuntu SMP Tue Aug 20 19:53:07 UTC 2019 x86_64 Linux`
Ruby: `2.5.3p105`
dry-validaiton: 1.3.1

Some other dry dependencies:
```
dry-schema (1.4.0)
dry-core (0.4.9)
```

Thanks in advance!

**Comment by ianwhite** [comment]
I wonder if this is to do with the multiple schemas inheritance introduced in #587 🤔?

I'll write a test case based on the above, and investigate.

**Comment by greyblake** [comment]
Thanks for the hint!
A workaround is for now is to lock `dry-schema` to version `1.3.4`.

**Comment by ianwhite** [comment]
I've just written a test case based on your example, and confirmed that it is a problem with dry-schema > 1.3.4.

**Comment by greyblake** [comment]
@ianwhite  thank you!

**Comment by ianwhite** [comment]
This should be fixed when dry-rb/dry-schema#196 is merged. That change has a test, so I don’t think we need the test above to be in dry-validation

**Comment by greyblake** [comment]
We've updated to dry-schema 1.4.1. Everything works smoothly:)
@ianwhite  @skryukov thank you guys!
